connection.go

```diff
@@ -1062,22 +1062,18 @@ func (conn *Connection) newFuture(ctx context.Context) (fut *Future) {
 }
 
 // This method removes a future from the internal queue if the context
-// is "done" before the response is come. Such select logic is inspired
-// from this thread: https://groups.google.com/g/golang-dev/c/jX4oQEls3uk
+// is "done" before the response is come.
 func (conn *Connection) contextWatchdog(fut *Future, ctx context.Context) {
 	select {
 	case <-fut.done:
+	case <-ctx.Done():
+	}
+
+	select {
+	case <-fut.done:
+		return
 	default:
-		select {
-		case <-ctx.Done():
-			conn.cancelFuture(fut, fmt.Errorf("context is done"))
-		default:
-			select {
-			case <-fut.done:
-			case <-ctx.Done():
-				conn.cancelFuture(fut, fmt.Errorf("context is done"))
-			}
-		}
```

Comment:
As you know, go doesn't ensure that if you have both channels done, the first in the select will be chosen. The given construction provides it. Also, this optimization allows us not to lock a mutex ahead of time. Are you sure of this change?

Reply:
> As you know, golang doesn't ensure that if you have both channels done, the first one in the select will be chosen. The given construction provides it.

1. The ordering is saved:

```Go
func (conn *Connection) contextWatchdog(fut *Future, ctx context.Context) {
        // Wait.
	select {
	case <-fut.done:
	case <-ctx.Done():
	}

        // Ordering.
	select {
	case <-fut.done:
		return
	default:
		conn.cancelFuture(fut, fmt.Errorf("context is done"))
	}
}
```

2. In fact it does not provide ordering in the old code:

```Go
			select {
			case <-fut.done:
			case <-ctx.Done():
				conn.cancelFuture(fut, fmt.Errorf("context is done"))
			}
```

> Also, this optimization allows us not to lock a mutex ahead of time. Are you sure of this change?

We're likely to have a lock here anyway (we just check req.Ctx().Done() and just send a request). So the optimization will not be helpful in this case.

Reply:
Thank you for explanation.

Reply:
Well it seems, that the first two check will never be successful unless we pass an already done (or near that) context, so let us leave your version.